In [ ]:
# !pip install -r requirements.txt

  Using cached aiohappyeyeballs-2.4.3-py3-none-any.whl (14 kB)
  Using cached aiohttp-3.10.10-cp310-cp310-macosx_11_0_arm64.whl (390 kB)
  Using cached aiosignal-1.3.1-py3-none-any.whl (7.6 kB)
  Using cached alembic-1.14.0-py3-none-any.whl (233 kB)
  Using cached annotated_types-0.7.0-py3-none-any.whl (13 kB)
  Using cached anthropic-0.39.0-py3-none-any.whl (198 kB)
  Using cached anyio-4.6.2.post1-py3-none-any.whl (90 kB)
  Using cached asgiref-3.8.1-py3-none-any.whl (23 kB)
  Using cached async_timeout-4.0.3-py3-none-any.whl (5.7 kB)
  Using cached attrs-24.2.0-py3-none-any.whl (63 kB)
  Using cached backoff-2.2.1-py3-none-any.whl (15 kB)
  Using cached bcrypt-4.2.0-cp39-abi3-macosx_10_12_universal2.whl (472 kB)
  Using cached beautifulsoup4-4.12.3-py3-none-any.whl (147 kB)
  Using cached build-1.2.2.post1-py3-none-any.whl (22 kB)
  Using cached cachetools-5.5.0-py3-none-any.whl (9.5 kB)
  Using cached certifi-2024.8.30-py3-none-any.whl (167 kB)
  Using cached charset_normalizer-3.4

In [2]:
import pandas as pd
med_qa_data = pd.read_csv('data/MedQA_complete_graded_data_encoded.csv')
med_qa_data = med_qa_data.drop(columns = ['index', 'Unnamed: 29', 'Unnamed: 30'])

In [3]:
med_qa_data.head()

question  \
0  Two weeks after undergoing an emergency cardia...   
1  A 46-year-old man is brought to the emergency ...   
2  A 30-year-old African American woman comes to ...   
3  A 67-year-old man who was diagnosed with arthr...   
4  A one-day-old male is evaluated in the hospita...   

                             answer                      new_IR IR Grade  \
0          Cholesterol embolization  Thrombotic microangiopathy        N   
1       Benzodiazepine intoxication        Alcohol intoxication        N   
2  Histoplasma capsulatum infection                Valley Fever        N   
3               Psoriatic arthritis         Psoriatic arthritis        Y   
4                  Duodenal atresia               Gastroschisis        N   

                        new_AR AR Grade                       new_BR BR Grade  \
0  Thrombotic microangiopathy.        N  Thrombotic microangiopathy.        N   
1         Alcohol intoxication        N         Alcohol intoxication        N   
2           Coccidioidomycosis        N               Histoplasmosis        Y   
3          Psoriatic arthritis        Y          Psoriatic arthritis        Y   
4       Intestinal obstruction        N                Gastroschisis        N   

                       new_DR DR Grade  ... DR Grade.1  \
0  Thrombotic microangiopathy        N  ...          Y   
1        Alcohol intoxication        N  ...          N   
2            Fungal infection        N  ...          Y   
3         Psoriatic arthritis        Y  ...          Y   
4            Duodenal atresia        Y  ...          Y   

                            IR  \
0  Atheroembolic renal disease   
1    Wernicke's encephalopathy   
2               Histoplasmosis   
3          Psoriatic Arthritis   
4               Meconium ileus   

                                         Rationale.2 IR Grade.1  \
0  This patient has decreased urinary output, mal...          Y   
1  This patient's symptoms of altered mental stat...          N   
2  This patient presents with fever, cough, chest...          Y   
3  The patient's clinical presentation suggests p...          Y   
4  This neonate's clinical presentation of biliou...          N   

                          AR  \
0       Cholesterol embolism   
1  Wernicke's encephalopathy   
2             Histoplasmosis   
3        Psoriatic Arthritis   
4         Intestinal atresia   

                                         Rationale.3 AR Grade.1  \
0  The patient's recent history of cardiac cathet...          Y   
1  The patient's symptoms (somnolence, slurred sp...          N   
2  The patient's history of recent hiking in Miss...          Y   
3  The patient presents with several features sug...          Y   
4  The symptoms presented by this neonate, such a...          Y   

                                  BR  \
0       Contrast-induced nephropathy   
1          Wernicke's encephalopathy   
2             Histoplasma capsulatum   
3                Psoriatic Arthritis   
4  Duodenal atresia in Down Syndrome   

                                         Rationale.4 BR Grade.1  
0  This patient's symptoms and clinical history s...          N  
1  The prior probability of Wernicke's encephalop...          N  
2  The prior probability of Histoplasmosis (a fun...          Y  
3  The prior probability of psoriatic arthritis i...          Y  
4  In Bayesian reasoning, we begin with a prior p...          Y  

[5 rows x 28 columns]

## Data cleaning and such

In [5]:
import re

# replace newline characters and " characters
med_qa_data['answer'] = med_qa_data['answer'].str.replace('\n', '')
med_qa_data['answer'] = med_qa_data['answer'].str.replace('"', '')

def remove_non_alphanumeric(input_string):
    if isinstance(input_string, str):
        return re.sub(r'[^a-zA-Z0-9 ]', ' ', input_string)
    return input_string

med_qa_data['answer'] = med_qa_data['answer'].apply(lambda row: remove_non_alphanumeric(row))

In [6]:
med_qa_data[med_qa_data['answer'].apply(lambda x: isinstance(x, str) and '"' in x)]

Empty DataFrame
Columns: [question, answer, new_IR, IR Grade, new_AR, AR Grade, new_BR, BR Grade, new_DR, DR Grade, CoT, CoT.1, GPT-4->, CoT.2, Rationale, CoT Grade, DR, Rationale.1, DR Grade.1, IR, Rationale.2, IR Grade.1, AR, Rationale.3, AR Grade.1, BR, Rationale.4, BR Grade.1]
Index: []

[0 rows x 28 columns]

In [7]:
med_qa_data.tail()

question  \
508  A previously healthy 21-year-old man comes to ...   
509  A 15-year-old boy is brought to the physician ...   
510  A 9-year-old boy is brought to a pediatric psy...   
511  A 28-year-old woman is brought to the emergenc...   
512                                                NaN   

                                       answer                     new_IR  \
508                  Chronic cerebral hypoxia  Carbon monoxide poisoning   
509                         Kallmann syndrome       Klinefelter Syndrome   
510  Attention deficit hyperactivity disorder                       ADHD   
511                         Bronchial rupture       Tension pneumothorax   
512                                       NaN                        NaN   

    IR Grade                 new_AR AR Grade                     new_BR  \
508        Y          Mold exposure        N  Carbon monoxide poisoning   
509        N        Delayed puberty        N           Delayed puberty.   
510        Y                   ADHD        Y                       ADHD   
511        N  Tension pneumothorax.        N       Tension pneumothorax   
512      248                    NaN      206                        NaN   

    BR Grade                new_DR DR Grade  ... DR Grade.1  \
508        Y                Anemia        N  ...          Y   
509        N       Delayed puberty        N  ...          N   
510        Y                  ADHD        Y  ...          Y   
511        N  Tension pneumothorax        N  ...          N   
512      216                   NaN      195  ...        385   

                                                    IR  \
508  This young man presents with headache, letharg...   
509  This patient's delayed puberty (Tanner stage 1...   
510  The child's distractibility, poor academic per...   
511  This patient's physical examination and the hi...   
512                                                NaN   

                               Rationale.2 IR Grade.1  \
508                         Radon exposure          N   
509                   Klinefelter Syndrome          N   
510  ADHD, predominantly inattentive type.          Y   
511                  Tension pneumothorax.          N   
512                                    NaN        379   

                                                    AR  \
508  This patient's symptoms, especially headache, ...   
509  This young male patient presents with a histor...   
510  This patient's inattentiveness, talkativeness,...   
511  This patient presents with evidence of severe ...   
512                                                NaN   

                                      Rationale.3 AR Grade.1  \
508         Chronic hypoxia-induced polycythemia.          Y   
509                          Klinefelter Syndrome          N   
510      Attention-Deficit/Hyperactivity Disorder          Y   
511  Traumatic Pneumothorax and Pneumomediastinum          N   
512                                           NaN        384   

                                                    BR  \
508  The starting point, or the prior probability, ...   
509  The prior probability of Klinefelter syndrome ...   
510  he prior probability of Attention-Deficit/Hype...   
511  The prior probability of a tension pneumothora...   
512                                                NaN   

                                   Rationale.4 BR Grade.1  
508          Chronic carbon monoxide exposure.          Y  
509                       Klinefelter Syndrome          N  
510  Attention-Deficit/Hyperactivity Disorder.          Y  
511                       Tension pneumothorax          N  
512                                        NaN        357  

[5 rows x 28 columns]

### We can see that that answers that were marked as correct don't *exactly* match the "golden answer" (adhd = "attention deficit hyperactivity disorder")

### There are total counts of the correct answers on the last row of the dataframe - Lets see if the value counts align with what they report in the dataframe

In [8]:
med_qa_data['IR Grade'].value_counts()

IR Grade
N      269
Y      243
248      1
Name: count, dtype: int64

##### In the Results section of the paper, the authors state that they achieved an accuracy of 78% using DR (differential diagnoses) with GPT 4 on 498 questions. This corresponds with the column 'DR Grade.1'

In [9]:
med_qa_data['DR Grade.1'].value_counts()

DR Grade.1
Y      388
N      108
y        1
385      1
Name: count, dtype: int64

In [10]:
389 / (388 + 1 + 108)

0.7826961770623743

#### Yay this matches the results that they got

In [11]:
389 + 108

497

Maybe they counted the result row by accident? Seems like real total is 497

In [12]:
med_qa_data.shape

(513, 28)

# The important columns to us are "question" and "answer" column
"answer" column is the golden answer

## Examine the questions that are asked

In [13]:
print(med_qa_data['question'].iloc[0])

Two weeks after undergoing an emergency cardiac catherization with stenting for unstable angina pectoris, a 61-year-old man has decreased urinary output and malaise. He has type 2 diabetes mellitus and osteoarthritis of the hips. Prior to admission, his medications were insulin and naproxen. He was also started on aspirin, clopidogrel, and metoprolol after the coronary intervention. His temperature is 38°C (100.4°F), pulse is 93/min, and blood pressure is 125/85 mm Hg. Examination shows mottled, reticulated purplish discoloration of the feet. Laboratory studies show:
Hemoglobin count 14 g/dL
Leukocyte count 16,400/mm3
Segmented neutrophils 56%
Eosinophils 11%
Lymphocytes 31%
Monocytes 2%
Platelet count 260,000/mm3
Erythrocyte sedimentation rate 68 mm/h
Serum
Urea nitrogen 25 mg/dL
Creatinine 4.2 mg/dL
Renal biopsy shows intravascular spindle-shaped vacuoles. What is the most likely cause of this patient's symptoms?"


In [14]:
print(med_qa_data['question'].iloc[40])

A 14-year-old boy is brought to the physician because of an increasing difficulty in hearing over the past several months. His mother says they have to speak at a higher volume for him to understand them. He also complains of having difficulty reading his favorite books because he is not able to see the words clearly. His father received a renal transplant in his 20s. The vital signs are within normal limits. The physical examination shows no abnormalities. Laboratory studies show:
Serum
Urea nitrogen 15 mg/dL
Creatinine 1.0 mg/dL
Urine
Blood 1+
Protein 1+
RBC 15–17/hpf
WBC 1–2/hpf
The audiometry shows bilateral high-frequency sensorineural hearing loss. The ophthalmologic examination shows anterior lenticonus. What best explains these findings?


In [15]:
med_qa_data.columns

Index(['question', 'answer', 'new_IR', 'IR Grade', 'new_AR', 'AR Grade',
       'new_BR', 'BR Grade', 'new_DR', 'DR Grade', 'CoT', 'CoT.1', 'GPT-4->',
       'CoT.2', 'Rationale', 'CoT Grade', 'DR', 'Rationale.1', 'DR Grade.1',
       'IR', 'Rationale.2', 'IR Grade.1', 'AR', 'Rationale.3', 'AR Grade.1',
       'BR', 'Rationale.4', 'BR Grade.1'],
      dtype='object')

In [22]:
med_qa_data.to_csv('data/cleaned_MedQA_complete_graded_data_encoded.csv', index= False)

In [209]:
nejm_df = pd.read_csv('data/NEJM_complete_graded.csv', encoding = "utf-8")

In [210]:
nejm_df.head()

DOI                                              Title  \
0  10.1056/NEJMcpc2300894  Case 18-2023: A 19-Year-Old Woman with Dyspnea...   
1  10.1056/NEJMcpc2201248  Case 17-2023: A 58-Year-Old Woman with Fatigue...   
2  10.1056/NEJMcpc2201237  Case 16-2023: A 13-Year-Old Boy with Depressio...   
3  10.1056/NEJMcpc2300895  Case 15-2023: A 33-Year-Old Man with Paresthes...   
4  10.1056/NEJMcpc2211511  Case 14-2023: A 31-Year-Old Man with Redness o...   

                                            Lab Data  \
0  Hemoglobin (g/dl)\t \t9.1\t \t8.4\nHematocrit ...   
1  Hemoglobin (g/dl)\t11.1–15.9\t11.7\t14.8...   
2  Hemoglobin (g/dl)\t \t12.8\nHematocrit (%)\t \...   
3                                                NaN   
4                                                NaN   

                                              Answer  \
0                        Takayasu’s arteritis.   
1                     Mansonella perstans infection.   
2                   Intentional guanfacine overdose.   
3  Functional vitamin B12 deficiency from use of ...   
4                               Mpox conjunctivitis.   

                                      Initial Answer  \
0                        Takayasu’s arteritis.   
1                      Mansonella perstans infection   
2  Toxic effects of a centrally acting _2-adrener...   
3  Functional vitamin B12 deficiency from use of ...   
4                               Mpox conjunctivitis.   

                                        CoT_response  \
0  {\n  "id": "chatcmpl-7hsJNxStzqjtgaemj7C5LGihk...   
1  {\n  "id": "chatcmpl-7hsKiZNYfdJrNnJMIFpi4PG4z...   
2  {\n  "id": "chatcmpl-7hsLzKL2SEHds78xQje5Uzl5x...   
3  {\n  "id": "chatcmpl-7hsN9myJ778yIgu6rUPNYXMU3...   
4  {\n  "id": "chatcmpl-7hsOUUAjRLFZ2nBSnTs5Hq2l9...   

                         CoT answer CoT Grade  \
0                Takayasu arteritis         Y   
1                   Schistosomiasis         N   
2             Adrenal insufficiency         N   
3  Nitrous oxide-induced neuropathy         Y   
4     Monkeypox with conjunctivitis         Y   

                                         DR_response  \
0  {\n  "id": "chatcmpl-7hsJwbroju1Ujb4YU8uuvr29v...   
1  {\n  "id": "chatcmpl-7hsLLyDbQxIvCSa7hJDd6kvB1...   
2  {\n  "id": "chatcmpl-7hsMVRwBMo1fCXiPLIyXnHff2...   
3  {\n  "id": "chatcmpl-7hsNn1p0nkQzA0MynPfTtoBpU...   
4  {\n  "id": "chatcmpl-7hsP8QXEvkq3XgBGSBByU2nvX...   

                                           DR answer DR grade  
0                                 Takayasu arteritis        Y  
1                                    Schistosomiasis        N  
2                                        Dehydration        N  
3  Nitrous oxide-induced subacute combined degene...        Y  
4                                          Monkeypox        Y

In [211]:
res_row = nejm_df.iloc[-1]

In [212]:
nejm_df = nejm_df[0:-1]

In [213]:
nejm_df.iloc[-1]

DOI                                          10.1056/NEJMcpc1415169
Title             Case 30-2015 — A 50-Year-Old Man with Cardioge...
Lab Data                                                        NaN
Answer            Myotonic dystrophy type 1, with associated car...
Initial Answer               Adult-onset myotonic dystrophy type 1.
CoT_response      {\n  "id": "chatcmpl-7i5HzXcISRH6UDVMYlGNYg4VL...
CoT answer                               Myotonic dystrophy type 1.
CoT Grade                                                         Y
DR_response       {\n  "id": "chatcmpl-7i5IadNWDQlPa12yVaVfmMls1...
DR answer                                Myotonic dystrophy type 1.
DR grade                                                          Y
Name: 299, dtype: object

In [214]:
nejm_df.iloc[4]['CoT_response']

'{\n  "id": "chatcmpl-7hsOUUAjRLFZ2nBSnTs5Hq2l9FykZ",\n  "object": "chat.completion",\n  "created": 1690691550,\n  "model": "gpt-4-0613",\n  "choices": [\n    {\n      "index": 0,\n      "message": {\n        "role": "assistant",\n        "content": "The patient initially presented with an itchy perianal rash and rectal pain. He was sexually active with multiple partners, which suggests a sexually transmitted infection. However, tests for common sexually transmitted infections such as HIV, gonorrhea, chlamydia, and syphilis were negative. The patient then developed a similar rash on his chest, abdomen, and back, along with fever, chills, fatigue, malaise, and headache. PCR testing of a swab specimen from the perianal vesicle detected the presence of nonvariola orthopoxvirus DNA, specifically mpox virus (MPXV; formerly known as monkeypox virus), clade II. The patient then developed redness and discomfort in his right eye, which became painful. Examination of the eye revealed conjunctiva

In [215]:
nejm_df[nejm_df['CoT_response'].isna()]

Empty DataFrame
Columns: [DOI, Title, Lab Data, Answer, Initial Answer, CoT_response, CoT answer, CoT Grade, DR_response, DR answer, DR grade]
Index: []

In [216]:
nejm_df['CoT_response'] = nejm_df['CoT_response'].apply(json.loads)
nejm_df['DR_response'] = nejm_df['DR_response'].apply(json.loads)

In [218]:
nejm_df['CoT_response'].iloc[4]

{'id': 'chatcmpl-7hsOUUAjRLFZ2nBSnTs5Hq2l9FykZ',
 'object': 'chat.completion',
 'created': 1690691550,
 'model': 'gpt-4-0613',
 'choices': [{'index': 0,
   'message': {'role': 'assistant',
    'content': 'The patient initially presented with an itchy perianal rash and rectal pain. He was sexually active with multiple partners, which suggests a sexually transmitted infection. However, tests for common sexually transmitted infections such as HIV, gonorrhea, chlamydia, and syphilis were negative. The patient then developed a similar rash on his chest, abdomen, and back, along with fever, chills, fatigue, malaise, and headache. PCR testing of a swab specimen from the perianal vesicle detected the presence of nonvariola orthopoxvirus DNA, specifically mpox virus (MPXV; formerly known as monkeypox virus), clade II. The patient then developed redness and discomfort in his right eye, which became painful. Examination of the eye revealed conjunctival injection and petechiae, but no abnormalitie

In [221]:
nejm_df.iloc[4]

DOI                                          10.1056/NEJMcpc2211511
Title             Case 14-2023: A 31-Year-Old Man with Redness o...
Lab Data                                                        NaN
Answer                                         Mpox conjunctivitis.
Initial Answer                                 Mpox conjunctivitis.
CoT_response      {'id': 'chatcmpl-7hsOUUAjRLFZ2nBSnTs5Hq2l9FykZ...
CoT answer                            Monkeypox with conjunctivitis
CoT Grade                                                         Y
DR_response       {'id': 'chatcmpl-7hsP8QXEvkq3XgBGSBByU2nvXJYAA...
DR answer                                                 Monkeypox
DR grade                                                          Y
Name: 4, dtype: object